ci/taos/plugins-good/pr-format-flawfinder.sh

```diff
@@ -0,0 +1,131 @@
+#
+# @see      https://dwheeler.com/flawfinder/
+# @see      https://sourceforge.net/projects/flawfinder/
+# @see      https://github.com/nnsuite/TAOS-CI
+# @author   Geunsik Lim <geunsik.lim@samsung.com>
+
+# @brief [MODULE] TAOS/pr-format-flawfinder
+function pr-format-flawfinder(){
+    echo "########################################################################################"
+    echo "[MODULE] TAOS/pr-format-flawfinder: Check security problems in source codes (*.c, *.cpp) with flawfinder"
+    pwd
+
+    # Check if server administrator install required commands
+    check_dependency flawfinder
+    check_dependency file
+    check_dependency grep
+    check_dependency cat
+    check_dependency wc
+    check_dependency awk
+
+    check_result="skip"
+
+    # Display the flawfinder version that is installed in the CI server.
+    # Note that the out-of-date version can generate an incorrect result.
+    flawfinder --version
+
+    # Read file names that a contributor modified(e.g., added, moved, deleted, and updated) from a last commit.
+    # Then, investigate C/C++ source code generated from *.patch files of the last commit.
+    FILELIST=`git show --pretty="format:" --name-only --diff-filter=AMRC`
+    for i in ${FILELIST}; do
+        # Skip obsolete folder
+        if [[ $i =~ ^obsolete/.* ]]; then
+            continue
+        fi
+        # Skip external folder
+        if [[ $i =~ ^external/.* ]]; then
+            continue
+        fi
+        # Handle only text files in case that there are lots of files in one commit.
+        echo "[DEBUG] file name is ( $i )."
+        if [[ `file $i | grep "ASCII text" | wc -l` -gt 0 ]]; then
+            # in case of source code files (*.c, *.cpp)
+            case $i in
+                # in case of C/C++ code
+                *.c|*.cpp)
+                    echo "[DEBUG] ( $i ) file is source code with the text format."
+                    static_analysis_sw="flawfinder"
+                    if [[ ! -z $pr_flawfinder_check_level ]]; then
+                        echo "[DEBUG] flawfinder: It's okay. The value of the flawfinder level is $pr_flawfinder_check_level."
+                        static_analysis_rules="--html --context --minlevel=$pr_flawfinder_check_level"
+                    else
+                        echo "[DEBUG] flawfinder: It's okay. The value of the flawfinder level (e.g. pr_flawfinder_check_level) is empty."
+                        echo "[DEBUG] flawfinder: You can declare the check level between 1(default) and 5 in the configuration file."
+                        echo "[DEBUG] flawfinder: The module executes a flaw inspection with level 1 by default."
+                        static_analysis_rules="--html --context"
+                    fi
+
+                    flawfinder_result="flawfinder_security_flaw_result"
+                    # Check C/C++ file.
+                    $static_analysis_sw $static_analysis_rules $i > ../report/${flawfinder_result}_${i}.html
```

Comment:
Does this make a log html file for each source file? Hmm..

Reply:
Good question, Please refer to  Line 98 "**break**". If a file of the commit includes a security flaw, the module reports "failure" status. Then the module does not try to inspect the next file by default as an intended policy.

Reply:
Yes. I have two suggestions about this.
1) `break` after inspection of each file: if a commit has some security flaws in multiple files, a user may need to make multiple PRs, which is redundant work.
2) too many html log files: if there is no error, `/report` folder contains too many log files, which means it's difficult for users to check other report logs (maybe, `/report/flawfinder/` folder would be another option).

So, my point it, how about merging these log files into a single report log? and check this merged file using `flawfinder`.

Reply:
> 2. Merge log files: if there is no error, /report folder contains too many log files,
The output structure has a rule consistently. For example, if the source file is nnstreamer-tensor.c, the output is ****_nnstreamer-tensor.txt file. if we try to merge lots of log file into the one file, the readability of the error report will be damaged.

Reply:
> 1. No break after inspection of each file: if a commit has some security flaws in multiple files, a user may need to make multiple force-push to check again and again, which is redundant work.

Right, However developers did not often update multiple log files at once in the past from our experiment result (e.g. CPPcheck module). So, we had changed the report style from "all report" to "step-by-step" style. :) If we meet new cases in the near future, let's customize this module again to give contributors efficiency and convenience.  (And we have been keeping 1pr-1commit rule well.)